### PR TITLE
Add Weighted Instruction Tuning loss to SFTTrainer

### LINKED ISSUE #6746: Add Weighted Instruction Tuning (WIT) loss to SFTTrainer
## Feature request

I would like to add [Weighted Instruction Tuning (WIT)](https://huggingface.co/papers/2507.07817) as an opt-in loss for `SFTTrainer`. I am one of the paper's authors. The original research implementation is available at https://github.com/kowndinya-renduchintala/WIT.

## Motivation

Conventional instruction tuning assigns zero loss to prompt tokens and unit loss to response tokens. WIT generalizes this by assigning independent weights `lambda_p` and `lambda_r` to prompt and response token losses:

`L_WIT = -sum(lambda_p * log p(prompt tokens) + lambda_r * log p(response tokens)) / N_nonzero`

where `N_nonzero` is the number of tokens whose corresponding prompt/response weight is non-zero. Thus:

- `(lambda_p, lambda_r) = (0, 1)` recovers conventional completion-only SFT.
- `(lambda_p, lambda_r) = (1, 1)` recovers full-sequence language modeling.

The TACL paper evaluates 525 SFT runs across five model configurations and three datasets, and reports that independently weighting prompt and response tokens improves generalization over conventional SFT across the studied settings.

## Proposed TRL API

Following the existing DFT integration through `SFTConfig.loss_type`:

```python
training_args = SFTConfig(
    loss_type="wit",
    prompt_loss_weight=0.2,
    completion_loss_weight=0.8,
)
```

The feature is opt-in. The proposed defaults for WIT-specific weights are `0.0` and `1.0`, matching conventional completion-only instruction tuning.

## Correctness and integration plan

- Preserve the prompt/completion boundary produced during prompt-completion preprocessing and construct per-token weights.
- Mask tokens whose configured weight is zero with `-100`, so Transformers' accumulated `num_items_in_batch` is exactly the WIT denominator across gradient-accumulation microbatches. When `average_tokens_across_devices=True`, current Transformers also aggregates this denominator across data-parallel ranks.
- Compute the weighted sum of shifted token NLLs and divide by `num_items_in_batch`; when it is unavailable (for direct loss calls), divide by the local count of non-zero-weight tokens.
- Support standard and conversational prompt-completion text datasets, padding-free batches, and packing.
- Reject incompatible modes clearly (language-modeling datasets, VLM datasets, assistant-only loss, formatting functions, Liger kernels, custom loss functions, and context/sequence parallelism) rather than silently changing semantics.
- Add unit tests for the formula, zero-weight masking, collation, preprocessing/packing, gradient-accumulation equivalence, validation errors, training, docs, and the paper index.

I searched existing TRL issues and PRs for "Weighted Instruction Tuning", "WIT", and "prompt loss weight" and did not find an existing implementation.

The implementation is available in #6747.

### PR DESCRIPTION
**Pull request by C-anwoy**
# What does this PR do?

Adds [Weighted Instruction Tuning (WIT)](https://huggingface.co/papers/2507.07817) as an opt-in `SFTTrainer` loss for text prompt-completion datasets:

```python
training_args = SFTConfig(
    loss_type="wit",
    prompt_loss_weight=0.2,
    completion_loss_weight=0.8,
)
```

WIT assigns independent weights to prompt and completion token losses. The weighted loss is normalized by the number of tokens whose weight is non-zero, rather than by the sum of weights. Consequently, `(0.0, 1.0)` recovers completion-only SFT and `(1.0, 1.0)` recovers full-sequence language modeling.

The implementation:

- preserves prompt/completion boundaries as per-token weights through preprocessing, truncation, packing, padding, and padding-free collation;
- masks zero-weight tokens in `labels`, allowing current Transformers to compute the exact nonzero-token denominator across gradient-accumulation microbatches;
- consumes `num_items_in_batch` in the custom loss so Trainer does not apply an additional gradient-accumulation division;
- supports standard and conversational text prompt-completion datasets;
- rejects unsupported combinations explicitly (language-modeling or vision datasets, assistant-only loss, formatting functions, Liger kernels, custom loss functions, and context/sequence parallelism);
- documents the API and adds the paper to TRL's paper index; and
- adds unit, preprocessing, collation, packing, gradient-accumulation, validation, and end-to-end training tests.

Validation performed locally on the current `main` branch:

- `make precommit`
- `python -m pytest -p no:rerunfailures tests/test_sft_trainer.py` — 116 passed, 166 skipped
- focused WIT selection — 18 passed

Fixes #6746

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? See #6746.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by C-anwoy** [comment]
Hi @qgallouedec, when you have a chance, could you please take a look at this PR or route it to the appropriate reviewer? 

I'm the co-first author of the WIT paper, which has been peer-reviewed and accepted to the Transactions of the Association for Computational Linguistics (TACL) -- the flagship journal of the ACL community with impact factor of 11.7 in 2025. I am available to explain the objective, validate the mathematical details, and address any requested changes. 

The PR adds WIT as an opt-in SFTTrainer loss, including exact nonzero-token normalization under gradient accumulation, compatibility validation, documentation, and tests. 

I have reviewed the code thoroughly from my side. 
Thank you!

**Comment by C-anwoy** [comment]
@qgallouedec @albertvillanova, when you have a chance, could you please take a look at this PR and approve it, or route it to the appropriate reviewer?
